test/Spec.hs

```diff
@@ -1,4 +1,18 @@
 module Main (main) where
 
+import Test.Hspec
+import Crocodealer.Config
+import Crocodealer.Core.Label
+
+
 main :: IO ()
-main = putStrLn ("Test suite not yet implemented" :: String)
+main = hspec $ do 
+    describe "Config.loadConfig" $ do
+        it "Should load the sample configuration correctly" $
+            loadConfig "example-config.toml" `shouldBe` (
+            pure $ Config {
```

Comment:
@chshersh  I have a type error here, could you help with that?

```
$ stack test
crocodealer> build (lib + exe + test)
Preprocessing library for crocodealer-0.0.0.0..
Building library for crocodealer-0.0.0.0..
Preprocessing test suite 'crocodealer-test' for crocodealer-0.0.0.0..
Building test suite 'crocodealer-test' for crocodealer-0.0.0.0..
[1 of 1] Compiling Main

/home/maex/projects/contrib/crocodealer/test/Spec.hs:12:13: error:
    • Ambiguous type variable ‘m0’ arising from a use of ‘loadConfig’
      prevents the constraint ‘(MonadIO m0)’ from being solved.
      Probable fix: use a type annotation to specify what ‘m0’ should be.
      These potential instances exist:
        instance [safe] MonadIO IO -- Defined in ‘Control.Monad.IO.Class’
        ...plus two instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    • In the first argument of ‘shouldBe’, namely
        ‘loadConfig "example-config.toml"’
      In the second argument of ‘($)’, namely
        ‘loadConfig "example-config.toml"
           `shouldBe`
             (pure
                $ Config
                    {configUsername = Just "my-user", configRepository = Nothing,
                     configIgnoredRepositories = [Repo "my-ignored-repo",
                                                  Repo "my-other-ignored-repo"],
                     configLabelRules = []})’
      In a stmt of a 'do' block:
        it "Should load the sample configuration correctly"
          $ loadConfig "example-config.toml"
              `shouldBe`
                (pure
                   $ Config
                       {configUsername = Just "my-user", configRepository = Nothing,
                        configIgnoredRepositories = [Repo "my-ignored-repo",
                                                     Repo "my-other-ignored-repo"],
                        configLabelRules = []})
   |
12 |             loadConfig "example-config.toml" `shouldBe` (
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

/home/maex/projects/contrib/crocodealer/test/Spec.hs:12:13: error:
    • Ambiguous type variable ‘m0’ arising from a use of ‘shouldBe’
      prevents the constraint ‘(Show (m0 Config))’ from being solved.
      Probable fix: use a type annotation to specify what ‘m0’ should be.
      These potential instances exist:
        instance Show a => Show (NonEmpty a) -- Defined in ‘GHC.Show’
        instance Show a => Show (Ratio a) -- Defined in ‘GHC.Real’
        instance Show a => Show (Set a)
          -- Defined in ‘containers-0.6.0.1:Data.Set.Internal’
        ...plus 19 others
        ...plus 31 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    • In the second argument of ‘($)’, namely
        ‘loadConfig "example-config.toml"
           `shouldBe`
             (pure
                $ Config
                    {configUsername = Just "my-user", configRepository = Nothing,
                     configIgnoredRepositories = [Repo "my-ignored-repo",
                                                  Repo "my-other-ignored-repo"],
                     configLabelRules = []})’
      In a stmt of a 'do' block:
        it "Should load the sample configuration correctly"
          $ loadConfig "example-config.toml"
              `shouldBe`
                (pure
                   $ Config
                       {configUsername = Just "my-user", configRepository = Nothing,
                        configIgnoredRepositories = [Repo "my-ignored-repo",
                                                     Repo "my-other-ignored-repo"],
                        configLabelRules = []})
      In the second argument of ‘($)’, namely
        ‘do it "Should load the sample configuration correctly"
              $ loadConfig "example-config.toml"
                  `shouldBe`
                    (pure
                       $ Config
                           {configUsername = Just "my-user", configRepository = Nothing,
                            configIgnoredRepositories = [Repo "my-ignored-repo", ....],
                            configLabelRules = []})’
   |
12 |             loadConfig "example-config.toml" `shouldBe` (
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

/home/maex/projects/contrib/crocodealer/test/Spec.hs:13:13: error:
    • Ambiguous type variable ‘m0’ arising from a use of ‘pure’
      prevents the constraint ‘(Applicative m0)’ from being solved.
      Probable fix: use a type annotation to specify what ‘m0’ should be.
      These potential instances exist:
        instance Applicative NonEmpty -- Defined in ‘GHC.Base’
        instance Applicative IO -- Defined in ‘GHC.Base’
        instance Applicative Maybe -- Defined in ‘GHC.Base’
        ...plus two others
        ...plus 13 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    • In the second argument of ‘shouldBe’, namely
        ‘(pure
            $ Config
                {configUsername = Just "my-user", configRepository = Nothing,
                 configIgnoredRepositories = [Repo "my-ignored-repo",
                                              Repo "my-other-ignored-repo"],
                 configLabelRules = []})’
      In the second argument of ‘($)’, namely
        ‘loadConfig "example-config.toml"
           `shouldBe`
             (pure
                $ Config
                    {configUsername = Just "my-user", configRepository = Nothing,
                     configIgnoredRepositories = [Repo "my-ignored-repo",
                                                  Repo "my-other-ignored-repo"],
                     configLabelRules = []})’
      In a stmt of a 'do' block:
        it "Should load the sample configuration correctly"
          $ loadConfig "example-config.toml"
              `shouldBe`
                (pure
                   $ Config
                       {configUsername = Just "my-user", configRepository = Nothing,
                        configIgnoredRepositories = [Repo "my-ignored-repo",
                                                     Repo "my-other-ignored-repo"],
                        configLabelRules = []})
   |
13 |             pure $ Config {
   |             ^^^^^^^^^^^^^^^...


--  While building package crocodealer-0.0.0.0 using:
      /home/maex/.stack/setup-exe-cache/x86_64-linux-tinfo6/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.5 --builddir=.stack-work/dist/x86_64-linux-tinfo6/Cabal-2.4.0.1 build lib:crocodealer exe:crocodealer test:crocodealer-test --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
Progress 1/2
```

Reply:
@mstruebing `hspec` doesn't support well calls to `IO` function from inside `describe`. You need to rewrite your code like this:

```haskell
main = do
    config <- loadConfig
    hspec configSpec
```

While we're on this, let's move config specs to a separate module `Test.Config` inside `test-suite` stanza. We probably will have more tests in the future so let's start introducing proper module hierarchy :slightly_smiling_face:

Reply:
I can't get it to work :/